### DISCUSSION TITLE
Relative glob excluding. I even don't know how to name it !!!

### DISCUSSION THREAD

**Discussion by ghost** [SCons/scons#4625] [Q&A]
Hello!

I have a problem and can't solve it without hacks.

The following code uses two different approaches. One uses `glob` function of `Dir` and works as I expect and the second uses global `Glob` which internally calls `Dir.glob` but works differently.

```python
import SCons

env = Environment()

_dir = "src/"

print("========== Dir ===========")
for f in env.Dir(_dir).glob("*.cpp", exclude="*.gen.cpp"):
	print(f)

print("========= Glob ===========")
for f in env.Glob(_dir + "*.cpp", exclude="*.gen.cpp"):
	print(f)
```

The output is following:

```
========== Dir ===========
src/f1.cpp
src/f2.cpp
========= Glob ===========
src/f1.cpp
src/f2.cpp
src/f3.gen.cpp
src/f4.gen.cpp
```

If I change the second example to prepend actual directory to exclude string it works.
`for f in env.Glob(_dir + "*.cpp", exclude=_dir + "*.gen.cpp"):`

So the problem is that I already receive string `src/*.cpp` and I need to create hacky invalid `Dir` `env.Dir("src/*.cpp").path_elements[-2].path` to get `src` string or to use `os.path.dirname("src/*.cpp")` which is more beautiful but still is a workaround.

Is there a way to force `env.Glob` work like it was called from actual `env.glob` or do you have any other ideas how to perform it without hacks or mabual filtering of returned list? (I can't change format in which I receive string)

**Comment by mwichmann** [comment]
You're right.  This is because the Node version of `Glob` uses the directory node of the directory in which we're executing the SConscript, while when you call `Dir().glob()` explicitly, you're setting the evaluation directory yourself, so everything happens in `src`.  The search for matches accounts for this, but the search for negative matches (excludes) does not - ends up doing the glob for the exclude patterns one level up, and thus not finding them. I'm not sure if this ends up considered a bug or not, but your approach of prepending the dir to the search pattern ends up being "right".
